v2/internal/runner/github.go

```diff
@@ -0,0 +1,174 @@
+package runner
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"github.com/projectdiscovery/gologger"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport/http"
+)
+
+type customTemplateRepo struct {
+	owner       string
+	reponame    string
+	gitCloneURL string
+}
+
+// This function download the custom template repository
+// scenario 1: -gtr custom-template.txt  flag has passed => Only download the repos. Do not update
+// scenario 2: -gtr custom-template.txt -tup github   => Update the repo(git pull)  and download if any new repo
+// Reason to add update and download logic in single function is scenario 2
+func (r *Runner) downloadCustomTemplates(ctx context.Context) {
+	downloadPath := filepath.Join(r.templatesConfig.TemplatesDirectory, customTemplateType)
+
+	for _, customTemplate := range r.customTemplates {
+		clonePath := customTemplate.getLocalRepoClonePath(downloadPath)
+
+		if !isDirectory(clonePath) {
+			err := customTemplate.cloneRepo(clonePath, r.options.GithubToken)
+			if err != nil {
+				gologger.Info().Msgf("%s", err)
+			} else {
+				gologger.Info().Msgf("Repo %s/%s cloned successfully at %s", customTemplate.owner, customTemplate.reponame, clonePath)
+			}
+			continue
+		}
+		if r.options.UpdateTemplates == customTemplateType {
+			err := customTemplate.pullChanges(clonePath, r.options.GithubToken)
+			if err == nil {
+				gologger.Info().Msgf("Repo %s/%s successfully pulled the changes.\n", customTemplate.owner, customTemplate.reponame)
+			} else {
+				gologger.Info().Msgf("%s", err)
+			}
+		}
+	}
+}
+
+// parseCustomTemplates function reads the options.GithubTemplateRepo list,
+// Checks the given repos are valid or not and stores them into runner.CustomTemplates
+func (r *Runner) parseCustomTemplates() {
+	gitHubClient := getGHClientIncognito()
+
+	for _, repoName := range r.options.GithubTemplateRepo {
+		owner, repo, err := getOwnerAndRepo(repoName)
+		if err != nil {
+			gologger.Info().Msgf("%s", err)
+			continue
+		}
+		githubRepo, err := getGithubRepo(gitHubClient, owner, repo)
+		if err != nil {
+			gologger.Info().Msgf("%s", err)
+			continue
+		}
+		customTemplateRepo := &customTemplateRepo{
+			owner:       owner,
+			reponame:    repo,
+			gitCloneURL: githubRepo.GetCloneURL(),
+		}
+		r.customTemplates = append(r.customTemplates, *customTemplateRepo)
+	}
+}
+
+// getOwnerAndRepo returns the owner, repo, err from the given string
+// eg. it takes input projectdiscovery/nuclei-templates and
+// returns owner=> projectdiscovery , repo => nuclei-templates
+func getOwnerAndRepo(reponame string) (owner string, repo string, err error) {
+	s := strings.Split(reponame, "/")
+	if len(s) != 2 {
+		err = errors.Errorf("wrong Repo name: %s", reponame)
+		return
+	}
+	owner = strings.Split(reponame, "/")[0]
+	repo = strings.Split(reponame, "/")[1]
```

Comment:
we can reuse the variable as we already performed the split:
`strings.Split(reponame, "/")` => `s`